File: usaco/section_3.1/stamps.cpp
/*
ID: huangch7
LANG: C++
TASK: stamps
*/

#include <fstream>
#include <vector>
#include <limits>

using namespace std;

int main() {
    ifstream fin("stamps.in");
    ofstream fout("stamps.out");

    int K, N;
    fin >> K >> N;

    vector<int> coins;
    vector<int> dp(2000010, numeric_limits<int>::max());
    dp[0] = 0;

    int coinValue;
    while (coins.size() < N) {
        fin >> coinValue;
        coins.push_back(coinValue);
    }

    int ans = 0;
    for (int i = 1; i < dp.size(); ++i) {
        int best = numeric_limits<int>::max();
        
        for (int j = 0; j < coins.size(); ++j) {
            int prev = i - coins[j];
            if (prev >= 0 && dp[prev] != numeric_limits<int>::max() && dp[prev] <= K - 1) {
                best = min(best, dp[prev] + 1);
            }
        }

        if (best == numeric_limits<int>::max()) {
            ans = i - 1;
            break;
        }
        dp[i] = best;
    }

    fout << ans << endl;
    fout.close();

    return 0;
}

File: usaco/section_3.1/stamps.py
"""
ID: huangch7
LANG: PYTHON3
TASK: stamps
"""

task_name = "stamps"

fin = open(f'{task_name}.in', 'r')
fout = open(f'{task_name}.out', 'w')

K, N = [int(x) for x in fin.readline().strip().split()]

coins = []
# dp[x] = y, values x need at least y coins
dp = [None] * 2000010
dp[0] = 0
while len(coins) < N:
    coins += [int(x) for x in fin.readline().strip().split()]

ans = 0
for i in range(1, len(dp)):
    best = None
    
    for j in range(len(coins)):
        prev = i - coins[j]
        if prev >= 0 and dp[prev] is not None and dp[prev] <= K - 1:
            if best is None or best > dp[prev] + 1:
                best = dp[prev] + 1
    if best is None:
        ans = i - 1        
        break
    dp[i] = best

# print(dp[:20])
# print(ans)
fout.write(f"{ans}\n")
fout.close()
